sdk/core/azure-core/src/main/java/com/azure/core/implementation/util/LruCache.java

```diff
@@ -0,0 +1,329 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+package com.azure.core.implementation.util;
+
+import com.azure.core.util.logging.ClientLogger;
+
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.locks.ReentrantLock;
+import java.util.function.Function;
+
+/**
+ * Least recently used cache that will retain the most recently used key-values until a cache size limit is reached.
+ * <p>
+ * The least recently used cache will eject values that have been added or accessed least recently, offering the ability
+ * to limit the size of the cache while still receiving the benefit of retaining the most recently used values. The
+ * cache does have limitations where it is possible for a most commonly used value to be ejected due to a streak of less
+ * commonly but more recently used values being added, which won't be covered in this implementation.
+ */
+public final class LruCache<Key, Value> {
+    private static final ClientLogger LOGGER = new ClientLogger(LruCache.class);
+
+    private final int maxCacheSize;
+
+    // Maintains the key-value pairs in the cache.
+    private final Map<Key, Node<Key, Value>> cache;
+
+    private final ReentrantLock lock = new ReentrantLock();
+
+    private Node<Key, Value> head;
+    private Node<Key, Value> tail;
```

Comment:
should these be volatile ?

Reply:
Yeah, they should, just wanted to get the base implementation in before needing to add AtomicReferenceUpdaters